torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -0,0 +1,380 @@
+        [](const tensorpipe::Error& error) {
+          if (error) {
+            LOG(WARNING) << "sending error response failed: " << error.what();
+            return;
+          }
+        });
+  }
+}
+
+void TensorPipeAgent::respond(std::shared_ptr<tensorpipe::Pipe>& pipe) {
+  pipeRead(
+      pipe,
+      [this, pipe](
+          const tensorpipe::Error& error, Message&& requestMessage) mutable {
+        // TODO: Handle server pipe read error
+        if (error) {
+          LOG(WARNING) << "Server read message: " << error.what();
+          return;
+        }
+
+        // Arm for next read
+        respond(pipe);
+
+        uint64_t messageId = requestMessage.id();
+
+        // Run the user function
+        std::shared_ptr<FutureMessage> futureResponseMessage;
+        try {
+          futureResponseMessage = cb_->operator()(requestMessage);
+        } catch (const std::exception& e) {
+          futureResponseMessage = std::make_shared<FutureMessage>();
+          futureResponseMessage->setError(e.what());
+        }
+
+        // Shortcut if immediately done
+        if (futureResponseMessage->completed()) {
+          sendCompletedResponseMessage(
+              pipe,
+              futureResponseMessage,
+              futureResponseMessage->error(),
+              messageId);
+        } else {
+          // Not complete yet
+          futureResponseMessage->addCallback(
+              [this, pipe, futureResponseMessage, messageId](
+                  const Message& /* unused */,
+                  const c10::optional<utils::FutureError>& err) mutable {
+                // Done
+                sendCompletedResponseMessage(
+                    pipe, futureResponseMessage, err, messageId);
+              });
+        }
+      });
+}
+
+std::shared_ptr<FutureMessage> TensorPipeAgent::send(
+    const WorkerInfo& toWorkerInfo,
+    Message&& requestMessage) {
+  TORCH_CHECK(
+      requestMessage.isRequest(),
+      "TensorPipeAgent::send(..) is only for sending requests.");
+
+  if (!isRunning_) {
+    auto err = c10::str(
+        "Node ",
+        RpcAgent::getWorkerInfo().id_,
+        "tried to send() a message of type ",
+        requestMessage.type(),
+        " but RPC is no longer running on this node.");
+    throw std::runtime_error(err);
+  }
+
+  const auto& url = findWorkerURL(toWorkerInfo);
+
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  // See if we already have a connection to this address or not
+  auto it = connectedPipes_.find(toWorkerInfo.id_);
+  if (it == connectedPipes_.end()) {
+    std::tie(it, std::ignore) = connectedPipes_.emplace(
+        toWorkerInfo.id_, tensorpipe::Pipe::create(context_, url));
+  }
+  std::shared_ptr<tensorpipe::Pipe>& pipe = it->second;
+
+  std::shared_ptr<FutureMessage> futureResponseMessage =
+      std::make_shared<FutureMessage>();
+  requestMessage.setId(nextMessageID_++);
+  pendingResponseMessage_[requestMessage.id()] = futureResponseMessage;
+
+  pipeWrite(
```

Comment:
I think the lock should be released before calling `pipeWrite`, because the callback be reentrant (i.e., be called inline), and it will try to acquire the lock, which might deadlock.

Reply:
We discussed in out 1:1 a bit. Successful messages would invoke callbacks from looping threads. Only error callbacks could be invoked inline. In this integration, we are not doing any retry on send() from error callback. So we shouldn't have deadlock?